### ISSUE TITLE
CI emits `bitsandbytes` UserWarning: inner dimension (8) is not aligned for fast kernel with blocksize=64

### ISSUE THREAD

**Issue by albertvillanova** [huggingface/trl#6580]
### What

CI emits new `UserWarning`s from `bitsandbytes` in the PEFT + quantization tests:

```
tests/test_kto_trainer.py::TestKTOTrainer::test_peft_with_quantization
tests/test_dpo_trainer.py::TestDPOTrainer::test_peft_with_quantization
tests/test_sft_trainer.py::TestSFTTrainer::test_peft_with_quantization
  /__w/trl/trl/.venv/lib/python3.14/site-packages/bitsandbytes/backends/cuda/ops.py:944: UserWarning: inner dimension (8) is not aligned for fast kernel with blocksize=64, falling back to slower implementation.
    warn(

tests/test_kto_trainer.py::TestKTOTrainer::test_peft_with_quantization
tests/test_dpo_trainer.py::TestDPOTrainer::test_peft_with_quantization
tests/test_sft_trainer.py::TestSFTTrainer::test_peft_with_quantization
  /__w/trl/trl/.venv/lib/python3.14/site-packages/bitsandbytes/backends/cuda/ops.py:944: UserWarning: inner dimension (32) is not aligned for fast kernel with blocksize=64, falling back to slower implementation.
    warn(
```

CI run: https://github.com/huggingface/trl/actions/runs/30421676482/job/90479609738

Affected tests:

- `tests/test_dpo_trainer.py::TestDPOTrainer::test_peft_with_quantization`
- `tests/test_kto_trainer.py::TestKTOTrainer::test_peft_with_quantization`
- `tests/test_sft_trainer.py::TestSFTTrainer::test_peft_with_quantization`

The tests themselves pass; for now this is warning noise only.

### When it started

The warnings appeared with the release of `bitsandbytes` 0.50.0 (2026-07-24), which CI now installs:

- https://pypi.org/project/bitsandbytes/0.50.0/
- https://github.com/bitsandbytes-foundation/bitsandbytes/releases/tag/0.50.0

Confirmed by running an equivalent 1-step SFT + LoRA + 4-bit (nf4) training on the same tiny model with each version: 0 such warnings with `bitsandbytes` 0.49.2, 26 with 0.50.0.

Same release that caused #6543.

### Next steps

Investigate the root cause, then decide whether anything is actionable on the TRL side and/or should be reported upstream.

**Comment by albertvillanova** [comment]
## Root cause pinned down

`bitsandbytes` 0.50.0 added a new `bitsandbytes::gemm_4bit` op (https://github.com/bitsandbytes-foundation/bitsandbytes/pull/1949, ROCm follow-up https://github.com/bitsandbytes-foundation/bitsandbytes/pull/1979) and moved the blocksize-alignment check into its dispatch — `bitsandbytes/backends/cuda/ops.py:944`, with the same warning at `bitsandbytes/backends/xpu/ops.py:189`:

```python
if M > _gemm_4bit_custom_max_m:
    use_custom = False
elif K % blocksize != 0:
    warn(
        f"inner dimension ({K}) is not aligned for fast kernel "
        f"with blocksize={blocksize}, falling back to slower implementation.",
        UserWarning,
    )
    use_custom = False
else:
    use_custom = _gemm_4bit_use_custom_fn(A.device.index, A.dtype, M, N, K)
```

`K` is the quantized layer's `in_features`. `trl-internal-testing/tiny-Qwen2ForCausalLM-2.5` has `hidden_size=8` and `intermediate_size=32`, so `K ∈ {8, 32}` — hence exactly the two messages we see: `K=8` for `q/k/v/o_proj` and `gate/up_proj`, `K=32` for `down_proj`. bitsandbytes accepts 4-bit blocksizes `(32, 64, 128, …, 4096)` and defaults to 64, and `transformers`' `BitsAndBytesConfig` exposes no blocksize option, so no configuration can make `K=8` aligned.

Confirmed the dimensions are the trigger: regenerating the same fixture with aligned dims (`hidden_size=64`, `intermediate_size=128`) and rerunning the identical job gives **0** warnings (26 → 0). Side observation for later: with aligned dims all 26 calls take the fused custom CUDA kernel instead of the dequant + `F.linear` fallback, and that choice is made by `_gemm_4bit_use_custom_cuda()`, which branches on compute capability and SM count (sm75/80/86/89/90/100/120, separate ROCm heuristic) plus `M`.

### Why it shows up only from 0.50.0, with no change on our side

Before 0.50.0 the same check lived in `matmul_4bit` (`bitsandbytes/autograd/_functions.py`) and was gated to the inference-only gemv path:

```python
if A.numel() == A.shape[-1] and A.requires_grad == False and A.device.type != "hpu":
    if A.shape[-1] % quant_state.blocksize != 0:
        warn(f"Some matrices hidden dimension is not a multiple of {quant_state.blocksize} ...")
```

i.e. single-row inputs (`M == 1`) that do not require grad. Training-time forwards all went to `MatMul4Bit.apply` (dequantize + matmul) silently. In 0.50.0, `matmul_4bit` routes **every** forward that does not need grad — `needs_grad = torch.is_grad_enabled() and (A.requires_grad or (bias is not None and bias.requires_grad))` — to the new `gemm_4bit` op, for all `M`. Training-time forwards therefore now reach the warning.

Measured on a single 1-step SFT + LoRA + nf4 run with the tiny fixture (26 emissions): every quantized `Linear4bit` of both layers is affected, and every emission comes from a forward with `torch.is_grad_enabled() == False` inside gradient checkpointing (turned on by PEFT's `prepare_model_for_kbit_training` for k-bit training). DPO and KTO add their no-grad reference-model forwards on top of that.

```
    1x  K= 32  model.layers.0.mlp.down_proj        grad_enabled=False in_checkpoint=True
    2x  K=  8  model.layers.0.mlp.gate_proj        grad_enabled=False in_checkpoint=True
    2x  K=  8  model.layers.0.mlp.up_proj          grad_enabled=False in_checkpoint=True
    2x  K=  8  model.layers.0.self_attn.k_proj     grad_enabled=False in_checkpoint=True
    2x  K=  8  model.layers.0.self_attn.o_proj     grad_enabled=False in_checkpoint=True
    2x  K=  8  model.layers.0.self_attn.q_proj     grad_enabled=False in_checkpoint=True
    2x  K=  8  model.layers.0.self_attn.v_proj     grad_enabled=False in_checkpoint=True
    ... same for model.layers.1 ...
```

### Nothing incorrect on the TRL side

- We only forward a user-supplied quantization config: `get_quantization_config()` (`trl/trainer/utils.py`) builds `BitsAndBytesConfig` from `ModelConfig` flags, and trainers pass `quantization_config` through to `from_pretrained`. We never select a blocksize (not exposed by transformers) and never select model dimensions.
- No `Linear4bit` module is created by TRL itself.
- The "slower implementation" bitsandbytes falls back to is dequantize + `F.linear`, which is exactly the path bitsandbytes ≤ 0.49.2 used for *all* 4-bit training. So 0.50.0 adds a message here, not a slowdown.

### This is not only about our tiny fixtures

Several widely used VLM checkpoints have a vision-tower `intermediate_size` that is not a multiple of 64, so their vision `down_proj` (`K = intermediate_size`) hits the same warning. Verified by quantizing weights of those exact shapes and calling `torch.ops.bitsandbytes.gemm_4bit` directly:

| checkpoint | vision `intermediate_size` | `% 64` | warns |
| --- | --- | --- | --- |
| `Qwen/Qwen2.5-VL-{3B,7B}-Instruct` | 3420 | 28 | yes |
| `Qwen/Qwen3-VL-8B-Instruct`, `HuggingFaceTB/SmolVLM(2)`, `google/siglip-so400m-patch14-384` (so also Idefics3 / PaliGemma-2 / Gemma-3 towers) | 4304 | 16 | yes |
| `Qwen/Qwen2.5-VL-7B-Instruct` text `down_proj` (control) | 11008 | 0 | no |

`transformers`' bitsandbytes quantizer excludes only `llm_int8_skip_modules` (unset by default), the tied/output embeddings and `_keep_in_fp32_modules` (`quantizers/base.py:234`), so in a default VLM QLoRA run the vision tower is quantized. The warning is emitted per forward, per misaligned layer, with no dedup and no `stacklevel`, so it points at bitsandbytes internals and names no module — for these checkpoints there is also nothing a user can do about it, since the dimension comes from the architecture.

### Next steps

- [x] Decide what to do on the TRL side.
- [x] Decide whether to report upstream, given that the noise is not specific to our test fixtures.

**Comment by albertvillanova** [comment]
## Upstream status

Reported upstream: https://github.com/bitsandbytes-foundation/bitsandbytes/issues/2027

**Not fixed yet.** Still present in the latest release (0.50.0) and in current `main` (checked `bitsandbytes-foundation/bitsandbytes@1f4007e`: the warning is at `bitsandbytes/backends/cuda/ops.py:948` and `bitsandbytes/backends/xpu/ops.py:190`, and the only commit touching those files since 0.50.0 is the unrelated #2023). Also reproduced with the rolling `continuous-release_main` wheel (`bitsandbytes-1.33.7.preview`, reporting `0.50.1.dev0`, published 2026-07-28): 10 emissions for 10 `gemm_4bit` calls on a misaligned shape.

What I asked for upstream, each with an existing precedent in their codebase:

1. emit the warning at most once per `(K, blocksize)` instead of once per call (`_cpu_paged_warned` style guard, or a `functools.cache`d helper as already used in that module);
2. pass `stacklevel=` so it points at the caller instead of `bitsandbytes/backends/cuda/ops.py`, and include the shape so the affected layer can be identified;
3. move the alignment check into the `else` branch, so it only warns when the fused kernel would otherwise have been selected (today it also warns where the arch/dtype heuristic would fall back anyway);
4. apply the same to the XPU copy.

I offered to send the PR for 1, 2 and 4.

Note the report is not only about our test fixtures: the vision towers of `Qwen/Qwen2.5-VL-3B/7B-Instruct` (`intermediate_size=3420`) and of the SigLIP-SO400M family (`4304`: `Qwen/Qwen3-VL-8B-Instruct`, `HuggingFaceTB/SmolVLM(2)`, and therefore Idefics3 / PaliGemma-2 / Gemma-3 towers) are misaligned as well, and transformers quantizes them by default, so QLoRA fine-tuning of those checkpoints hits the same warning with no way for the user to avoid it.

Still to decide: what we do on the TRL side in the meantime.
